# Tugas UTS

Muhammad Adam Zaky Jiddyansah

210411100234

Pada Tugas UTS ini berupa menyelesaikan soal yang tertera pada LMS schology Penambangan Data, soal sebagai berikut :

1. Lakukan analisa data klasifikasi dengan sumber data https://github.com/kb22/Heart-Disease-Prediction/blob/master/dataset.csv dengan menggunakan naive bayes dan knn untuk data  
2. Jelakan proses tersebut dengan menulisnya di blog github yang anda telah miliki ( jupyter book).
3. Upload alamat blog github anda di schoology

## Dengan Metode KNN

In [146]:
import pandas as pd
import numpy as np

In [147]:
data = pd.read_csv('https://raw.githubusercontent.com/kb22/Heart-Disease-Prediction/master/dataset.csv')
data

age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  \
0     63    1   3       145   233    1        0      150      0      2.3   
1     37    1   2       130   250    0        1      187      0      3.5   
2     41    0   1       130   204    0        0      172      0      1.4   
3     56    1   1       120   236    0        1      178      0      0.8   
4     57    0   0       120   354    0        1      163      1      0.6   
..   ...  ...  ..       ...   ...  ...      ...      ...    ...      ...   
298   57    0   0       140   241    0        1      123      1      0.2   
299   45    1   3       110   264    0        1      132      0      1.2   
300   68    1   0       144   193    1        1      141      0      3.4   
301   57    1   0       130   131    0        1      115      1      1.2   
302   57    0   1       130   236    0        0      174      0      0.0   

     slope  ca  thal  target  
0        0   0     1       1  
1        0   0     2       1  
2        2   0     2       1  
3        2   0     2       1  
4        2   0     2       1  
..     ...  ..   ...     ...  
298      1   0     3       0  
299      1   0     3       0  
300      1   2     3       0  
301      1   1     3       0  
302      1   1     2       0  

[303 rows x 14 columns]

### Eksplorasi Data

In [148]:
data.head()


age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  slope  \
0   63    1   3       145   233    1        0      150      0      2.3      0   
1   37    1   2       130   250    0        1      187      0      3.5      0   
2   41    0   1       130   204    0        0      172      0      1.4      2   
3   56    1   1       120   236    0        1      178      0      0.8      2   
4   57    0   0       120   354    0        1      163      1      0.6      2   

   ca  thal  target  
0   0     1       1  
1   0     2       1  
2   0     2       1  
3   0     2       1  
4   0     2       1

In [149]:
data.shape


(303, 14)

### Memindahkan dataset menjadi input(X) dan target (Y)

Selanjutnta membuat dataframe dengan semua training data kecuali kolom target

In [150]:
X = data.drop(columns=["target"])

Lalu mengecek apakah variabel target sudah dihapus


In [151]:
X.head()


age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  slope  \
0   63    1   3       145   233    1        0      150      0      2.3      0   
1   37    1   2       130   250    0        1      187      0      3.5      0   
2   41    0   1       130   204    0        0      172      0      1.4      2   
3   56    1   1       120   236    0        1      178      0      0.8      2   
4   57    0   0       120   354    0        1      163      1      0.6      2   

   ca  thal  
0   0     1  
1   0     2  
2   0     2  
3   0     2  
4   0     2

In [152]:
Y = data["target"].values
Y[0:150]

array([1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1])

Memisahkan dataset menjadi train dan test data dan Scaling Data

In [153]:
from sklearn.model_selection import train_test_split
X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size=0.2, random_state=1, stratify=Y)

In [154]:
from sklearn.preprocessing import MinMaxScaler
scaler = MinMaxScaler()
scaler.fit(X_train)

X_train_scaled = scaler.transform(X_train)
X_test_scaled = scaler.transform(X_test)

Lalu Create dan Testing Model. serta membuat KNN Klasifikasi

In [155]:
from sklearn.neighbors import KNeighborsClassifier
knn = KNeighborsClassifier(n_neighbors = 3)
knn.fit(X_train_scaled,Y_train)

KNeighborsClassifier(n_neighbors=3)

### Menampilkan 5 model prediksi pertama pada data test

In [156]:
knn.predict(X_test_scaled)[0:5]

array([1, 1, 1, 1, 1])

Cek akurasi pada model test data

In [157]:
knn.score(X_test_scaled, Y_test)

0.8032786885245902

### Kesimpulan


*   Data Heart Disease Prediction terdiri dari 302 baris data.
*   Di data Heart Disease Prediction ini mempunyai ukuran 303, 14 dapat dilihat dengan df.shape
*   Melakukan pembagian dataset menjadi train set dan test set
*   Sebelum menentukan hasil akurasi perlu membuat testing model



## Naive Bayes

### Pertama Import beberapa modul python yang diperlukan





In [158]:
import pandas as pd
import numpy as np

In [159]:
datasetscr = "https://raw.githubusercontent.com/kb22/Heart-Disease-Prediction/master/dataset.csv"

In [160]:
df = pd.read_csv(datasetscr)
df.head()

age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  slope  \
0   63    1   3       145   233    1        0      150      0      2.3      0   
1   37    1   2       130   250    0        1      187      0      3.5      0   
2   41    0   1       130   204    0        0      172      0      1.4      2   
3   56    1   1       120   236    0        1      178      0      0.8      2   
4   57    0   0       120   354    0        1      163      1      0.6      2   

   ca  thal  target  
0   0     1       1  
1   0     2       1  
2   0     2       1  
3   0     2       1  
4   0     2       1

cek jumlah baris dan kolom yang ada di dataset

In [161]:
df.shape


(303, 14)

### Menghitung Data

Pada Langkah ini pisahkan kolom "target" dari data frame dan ambil kolom "target" sebagai target kolom untuk mengkategorikan

Selanjutnya menentukan variabel independen dan variabel dependen dari data yang akan dianalisis

In [162]:
x = df.drop(['target'], axis = 1)
x.head()

age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  slope  \
0   63    1   3       145   233    1        0      150      0      2.3      0   
1   37    1   2       130   250    0        1      187      0      3.5      0   
2   41    0   1       130   204    0        0      172      0      1.4      2   
3   56    1   1       120   236    0        1      178      0      0.8      2   
4   57    0   0       120   354    0        1      163      1      0.6      2   

   ca  thal  
0   0     1  
1   0     2  
2   0     2  
3   0     2  
4   0     2

Pada Kolom result of treatment di drop atau di hapus dari data frame karena akan menjadi variabel dependen

In [163]:
y = df['target']
y.head()

0    1
1    1
2    1
3    1
4    1
Name: target, dtype: int64

Selanjutnta melakukan analisis menggunakan klasifikasi naive bayes. untuk langkah dengan membagi dataset menjadi training set dan test set

In [164]:
# Import train_test_split function
from sklearn.model_selection import train_test_split
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size = 0.2, random_state = 123)

Di program atas saya membagi dataset menjadi 80% training set dan 20% test set

Lalu memisahkan untuk melakukan prediksi pada training set dan test set

In [165]:
# Import Gaussian Naive Bayes model
from sklearn.naive_bayes import GaussianNB
# Mengaktifkan/memanggil/membuat fungsi klasifikasi Naive bayes
modelnb = GaussianNB()
# Memasukkan data training pada fungsi klasifikasi naive bayes
nbtrain = modelnb.fit(x_train, y_train)
nbtrain.class_count_

array([108., 134.])

Menentukan hasil prediksi dari x_test

In [166]:
y_pred = nbtrain.predict(x_test)
y_pred

array([1, 0, 0, 0, 1, 0, 1, 0, 1, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0,
       1, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1,
       1, 0, 0, 0, 1, 1, 0, 0, 1, 0, 1, 1, 0, 1, 0, 1, 0])

Menentukan probbabilitas prediksi

In [167]:
nbtrain.predict_proba(x_test)


array([[5.82571613e-04, 9.99417428e-01],
       [9.99240610e-01, 7.59390366e-04],
       [9.99938869e-01, 6.11309321e-05],
       [9.99992658e-01, 7.34177465e-06],
       [3.63366729e-03, 9.96366333e-01],
       [9.99951069e-01, 4.89308903e-05],
       [8.58720034e-03, 9.91412800e-01],
       [5.98819913e-01, 4.01180087e-01],
       [5.96228807e-04, 9.99403771e-01],
       [9.33511439e-01, 6.64885614e-02],
       [9.97942318e-01, 2.05768198e-03],
       [4.57117589e-01, 5.42882411e-01],
       [9.99999986e-01, 1.36224796e-08],
       [9.99995730e-01, 4.26988712e-06],
       [5.16991611e-01, 4.83008389e-01],
       [1.00000000e+00, 5.08353833e-11],
       [9.10174530e-01, 8.98254697e-02],
       [9.86808541e-01, 1.31914589e-02],
       [5.72076923e-01, 4.27923077e-01],
       [2.24348439e-01, 7.75651561e-01],
       [1.35283843e-01, 8.64716157e-01],
       [9.99163302e-01, 8.36698360e-04],
       [4.11827050e-05, 9.99958817e-01],
       [9.82142245e-01, 1.78577552e-02],
       [9.874878

Setelah memperoleh hasil nilai prediksi y_pred, tahapan lanjutnya adalah melakukan Confusion Matrix

In [168]:
from sklearn.metrics import confusion_matrix
confusion_matrix(y_test, y_pred)

array([[27,  3],
       [ 7, 24]])

Langkah terakhir yaitu melakukan perhitungan hasil nilai akurasi

In [169]:
from sklearn.metrics import classification_report
print(classification_report(y_test,y_pred))

              precision    recall  f1-score   support

           0       0.79      0.90      0.84        30
           1       0.89      0.77      0.83        31

    accuracy                           0.84        61
   macro avg       0.84      0.84      0.84        61
weighted avg       0.84      0.84      0.84        61



### Kesimpulan
Setelah mengklasifikasi dengan membagi dataset menjadi 2 yaitu training set dan test set, maka akan dihitung hasil prediksi dengan Confusion Matriks agar dapat melakukan perhitungan Hasil Nilai Akurasi.

Maka diperoleh hasil akurasi ***0,84***